<!-- -*- mode: XML -*- -->

<launch>
    <!-- Start the Dynamixel servo manager to control wubble arm -->
    <node name="dynamixel_manager_arm" pkg="dynamixel_controllers" type="controller_manager.py" required="true" output="screen">
        <rosparam>
            namespace: wubble_arm_manager
            serial_ports:
                wubble2UpperArm:
                    port_name: "/dev/wubble2UpperArm"
                    baud_rate: 1000000
                    min_motor_id: 1
                    max_motor_id: 4
                    update_rate: 50
                wubble2LowerArm:
                    port_name: "/dev/wubble2LowerArm"
                    baud_rate: 1000000
                    min_motor_id: 5
                    max_motor_id: 9
                    update_rate: 50
        </rosparam>
    </node>

    <!-- Start upper and lower arm joint controllers -->
    <rosparam file="$(find wubble2_robot)/config/wubble_arm.yaml" command="load"/>
    <node name="upper_arm_controller_spawner" pkg="dynamixel_controllers" type="controller_spawner.py"
          args="--manager=wubble_arm_manager
                --port=wubble2UpperArm
                --type=simple
                shoulder_pitch_controller
                shoulder_pan_controller
                upperarm_roll_controller
                elbow_flex_controller"
          output="screen"/>
    <node name="lower_arm_controller_spawner" pkg="dynamixel_controllers" type="controller_spawner.py"
          args="--manager=wubble_arm_manager
                --port=wubble2LowerArm
                --type=simple
                forearm_roll_controller
                wrist_pitch_controller
                wrist_roll_controller"
          output="screen"/>

    <!-- Start gripper joint controllers -->
    <rosparam file="$(find wubble2_robot)/config/wubble_gripper.yaml" command="load"/>
    <node name="gripper_controller_spawner" pkg="dynamixel_controllers" type="controller_spawner.py"
          args="--manager=wubble_arm_manager
                --port=wubble2LowerArm
                --type=simple
                left_finger_controller
                right_finger_controller"
          output="screen"/>

    <!-- Start joint trajectory action controller -->
    <rosparam file="$(find wubble2_robot)/config/wubble_arm_trajectory_controller.yaml" command="load"/>
    <node name="trajectory_controller_spawner" pkg="dynamixel_controllers" type="controller_spawner.py"
          args="--manager=wubble_arm_manager
                --type=meta
                l_arm_controller
                shoulder_pitch_controller
                shoulder_pan_controller
                upperarm_roll_controller
                elbow_flex_controller
                forearm_roll_controller
                wrist_pitch_controller
                wrist_roll_controller"
          output="screen"/>

    <!-- Start gripper action controller -->
    <node name="gripper_action_controller" pkg="wubble2_robot" type="wubble_gripper_action_controller.py" output="screen" />

    <!-- Start gripper grasp action controller -->
    <node name="gripper_grasp_controller" pkg="wubble2_robot" type="wubble_gripper_grasp_controller.py" output="screen" >
        <param name="object_presence_pressure_threshold" value="1073.50"/>
        <param name="object_presence_opening_threshold" value="0.026"/>
        <param name="gripper_action_name" value="/wubble_gripper_action"/>
        <param name="posture_action_name" value="/wubble_gripper_grasp_action"/>
        <param name="grasp_query_name" value="/wubble_grasp_status"/>
    </node>
</launch>

